### PR TITLE
Replace `transfer` (deprecated) occurrences with `call` in smt tests

### PR DESCRIPTION
**Pull request by matheusaaguiar**
`transfer` is deprecated and will be removed in the next breaking release.
~depends on #16174.~ Merged.

### CODE REVIEW

File: test/cmdlineTests/model_checker_targets_all_all_engines/err [hunk 1]
```diff
@@ -54,48 +51,52 @@ Counterexample:
...
 x = 0
+success = true
 
 Transaction trace:
 test.constructor()
 State: arr = []
 test.f(0x0, 1)
-  --> input.sol:13:3:
+    a.call{value: x}("") -- untrusted external call
+  --> input.sol:14:3:
```
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_all_all_engines/err]: Hmm... so `transfer()` was not treated as an untrusted external call but `call()` is? Perhaps SMTChecker should treat `call()` with empty payload and a gas limit specially?

File: test/cmdlineTests/model_checker_targets_all_bmc/err [hunk 2]
```diff
@@ -45,31 +43,23 @@ Warning: BMC: Division by zero happens here.
...
-10 | 		a.transfer(x);
-   | 		^^^^^^^^^^^^^
-Note: Counterexample:
-  a = 0
-  x = 0
-
-Note: Callstack:
-Note:
+Note: 
+Note that external function calls are not inlined, even if the source code of the function is available. This is due to the possibility that the actual called contract has the same ABI but implements the function differently.
```
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_all_bmc/err]: Apparently SMTChecker can figure this out with `transfer()` but not with `call()`?

File: test/cmdlineTests/model_checker_targets_balance_bmc/err
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_balance_bmc/err:1]: Looks like SMTChecker can no longer detect insufficient balance when we use `call()`. Since this is the point of the test, we should replace it with something that actually does trigger this.

Comment (cameel) [reply test/cmdlineTests/model_checker_targets_balance_bmc/err:1]: Alternatively, we could also keep `transfer()` in this particular test. There's a chance we'll drop BMC before we drop `transfer()` and `send()` anyway.

File: test/cmdlineTests/model_checker_targets_balance_chc/err
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_balance_chc/err:1]: Same here, but we're not dropping CHC, so this one needs an actual fix.

File: test/cmdlineTests/model_checker_targets_constant_condition_chc/err
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_constant_condition_chc/err:1]: Looks like SMTChecker fails to detect the constant condition here (though it was like this already before this PR).

File: test/cmdlineTests/model_checker_targets_out_of_bounds_bmc/err
Comment (cameel) [inline test/cmdlineTests/model_checker_targets_out_of_bounds_bmc/err:1]: Another case where the target is not actually reported.

File: test/cmdlineTests/standard_model_checker_targets_assert_bmc/output.json [hunk 3]
```diff
@@ -2,66 +2,10 @@
...
-        }
-    ],
-    "sources": {
-        "A": {
-            "id": 0
+            "formattedMessage": "parse error at line 7, column 458: syntax error while parsing object - unexpected string literal; expected '}'",
+            "message": "parse error at line 7, column 458: syntax error while parsing object - unexpected string literal; expected '}'",
+            "severity": "error",
+            "type": "JSONError"
         }
```
Comment (cameel) [inline test/cmdlineTests/standard_model_checker_targets_assert_bmc/output.json:65]: This is not the expected output here. I think you broke it by not escaping the quotes in JSON.

File: test/libsolidity/smtCheckerTests/blockchain_state/transfer_2.sol [hunk 4]
```diff
@@ -4,13 +4,12 @@ contract C {
...
+		(success, ) = recipient.call{value: amount}("");
+		require(success);
 	}
 }
 // ====
 // SMTEngine: chc
 // ----
-// Warning 9207: (133-151): 'transfer' is deprecated and scheduled for removal in the next breaking version (0.9). Use 'call{value: <amount>}("")' instead.
-// Warning 9207: (167-185): 'transfer' is deprecated and scheduled for removal in the next breaking version (0.9). Use 'call{value: <amount>}("")' instead.
-// Info 1391: CHC: 2 verification condition(s) proved safe! Enable the model checker option "show proved safe" to see all of them.
```
Comment (cameel) [inline test/libsolidity/smtCheckerTests/blockchain_state/transfer_2.sol:16]: This test no longer is proved safe with `call()`.

**Review by cameel — COMMENTED** [review]
This needs a look from @blishko.

Looks like some of the balance tests actually do depend on properties of `transfer()` and SMTChecker is not able to prove the same things with `call()`. We'd need something else to replace it.

Also, some of the CLI tests for specific targets have empty output for those targets, which means we're not actually testing that the things works.

I'm not sure how much of it is actually worth fixing. We plan to deprecate BMC in 0.9 anyway. Some of these apply to CMC though.

**Comment by blishko** [comment]
How important is this right now?
I guess we need to implement handling of `call` with value the same way we now handle `transfer`?

**Comment by cameel** [comment]
> How important is this right now?

We will have to deal with it in some way eventually, because `send()`/`transfer()` are going to be removed and these tests will start failing, but it will only happen in 0.9, so it's not urgent at all. It only came up now because in #16174 we're adding deprecation warnings and it seemed straightforward to update these tests. Looks like it's not that straightforward after all :)

> I guess we need to implement handling of `call` with value the same way we now handle `transfer`?

Another option is also to wait for the PAY opcode and replace it with that. If it gets into Glamsterdam it could even happen before 0.9 ships :)

Maybe just leaving it as is and going back to it when the bultins are actually removed is the best course of action here. But I was just wondering whether you'd still plan to address this properly or if we should rather assume that this will just become irrelevant at some point due to changing plans (deprecations, move to Core Solidity, new tool for verification at Yul level, etc.).

**Comment by blishko** [comment]
I believe it will eventually become irrelevant, but we might need to fix it properly before that.
I can do it before 0.9, but I would prefer to do all the necessary things in one batch.

So I would wait with this a bit to see if the situation is clearer in the future (e.g., the PAY opcode).

**Comment by cameel** [comment]
ok. In that case in this PR we'll just replace the simple cases that cause no issues and leave the ones that do (e.g. the ones that are related to balance or specifically test `transfer()`) unchanged, to be dealt with later.

**Comment by matheusaaguiar** [comment]
I have removed the problematic tests. @cameel

**Comment by blishko** [comment]
It's not great that `call` is now treated as "untrusted external call".
If this is really equivalent to `transfer` that I should go and handle this special case of `call` the same way `transfer` is now handled.
Hopefully, you wouldn't need the `require(success)` anymore in that case as well.

**Comment by blishko** [comment]
I created an issue for myself: https://github.com/argotorg/solidity/issues/16400

In the current SMTChecker code, `call` is handled as general untrusted external call, while `transfer/send` has special treatment.

My current opinion is that even `transfer/send` should have been treated as untrusted external call, because they possibly allow reentrancy. So the fact that `call` is marked as such in the counterexamples is a good thing actually!
What is missing is creating a verification target in case `call` is sending non-zero value.

**Comment by blishko** [comment]
I think this PR can be merged as is now.
I will try to get to the issue at some point, so that we can update all the other tests as well.

**Review by blishko — APPROVED** [review]

**Review by cameel — APPROVED** [review]

**Comment by cameel** [comment]
> If this is really equivalent to `transfer` that I should go and handle this special case of `call` the same way `transfer` is now handled.

Not every `call()` though. Only the ones with a low enough gas limit.
